src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,407 @@
+            // Startup transient: queue hasn't gathered numHits yet
+            int slot = hitsCollected - 1;
+            // Copy hit into queue
+            if (numHits > 0) {
+                comparators[subQueryNumber].copy(slot, doc);
+                add(slot, doc, compoundScores[subQueryNumber], subQueryNumber, score);
+                if (queueFull[subQueryNumber]) {
+                    comparators[subQueryNumber].setBottom(bottom.slot);
+                }
+            } else {
+                queueFull[subQueryNumber] = true;
+            }
+        }
+
+        /*
+         * This hit is competitive - replace bottom element in queue & adjustTop
+         */
+        protected void collectCompetitiveHit(int doc, int subQueryNumber) throws IOException {
+            // This hit is competitive - replace bottom element in queue & adjustTop
+            if (numHits > 0) {
+                comparators[subQueryNumber].copy(bottom.slot, doc);
+                updateBottom(doc, compoundScores[subQueryNumber]);
+                comparators[subQueryNumber].setBottom(bottom.slot);
+            }
+        }
+
+        protected boolean thresholdCheck(int doc, int subQueryNumber) throws IOException {
+            if (collectedAllCompetitiveHits || reverseMul * comparators[subQueryNumber].compareBottom(doc) <= 0) {
+                // since docs are visited in doc Id order, if compare is 0, it means
+                // this document is larger than anything else in the queue, and
+                // therefore not competitive.
+                if (searchSortPartOfIndexSort) {
+                    if (hitsThresholdChecker.isThresholdReached()) {
+                        setTotalHitsRelation(TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO);
+                        log.info("Terminating collection as hits threshold is reached");
+                        throw new CollectionTerminatedException();
+                    } else {
+                        collectedAllCompetitiveHits = true;
+                    }
+                }
+                return true;
+            }
+            return false;
+        }
+
+        /*
+        The method initializes once per search request.
+         */
+        protected void initializePriorityQueuesWithComparators(LeafReaderContext context, int numberOfSubQueries) throws IOException {
+            if (compoundScores == null) {
+                compoundScores = new FieldValueHitQueue[numberOfSubQueries];
+                comparators = new LeafFieldComparator[numberOfSubQueries];
+                queueFull = new boolean[numberOfSubQueries];
+                collectedHits = new int[numberOfSubQueries];
+                for (int i = 0; i < numberOfSubQueries; i++) {
+                    initializeLeafFieldComparators(context, i);
+                }
+            }
+            if (isComparatorsInitialized) {
```

Comment:
shouldn't it be the other way around - check and initialize comparators if they are not yet initialized? Either the logic is wrong or variable names are not descriptive

Reply:
No, it does not work that ways.
1. The code which is under the flag **isComparatorsInitialized** needs to be executed for every segment.
2. HybridTopDocSortLeafCollector constructor gets initialized for every segment. Therefore, comparators need to be refreshed for every segment.

The reason why we need this flag is because we need to lazy init this.

Reference:
https://github.com/apache/lucene/blob/7e31f56ea1130909948bcf9e5beefc161ceef137/lucene/core/src/java/org/apache/lucene/search/TopFieldCollector.java#L52